src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -865,50 +950,99 @@ private async Task<Dictionary<KeyValueIdentifier, ConfigurationSetting>> LoadKey
+
+            var existingKeys = new HashSet<string>();
+
+            AsyncPageable<ConfigurationSetting> pageableSettings;
+
+            foreach (KeyValueSelector loadOption in selectors)
             {
-                IEnumerable<ConfigurationSetting> currentKeyValues = GetCurrentKeyValueCollection(changeWatcher.Key, changeWatcher.Label, existingSettings.Values);
+                if (string.IsNullOrEmpty(loadOption.SnapshotName))
+                {
+                    SettingSelector selector = new SettingSelector()
+                    {
+                        KeyFilter = loadOption.KeyFilter,
+                        LabelFilter = loadOption.LabelFilter ?? LabelFilter.Null
+                    };
+
+                    pageableSettings = client.GetConfigurationSettingsAsync(selector, cancellationToken);
+                }
+                else
+                {
+                    ConfigurationSnapshot snapshot;
+
+                    try
+                    {
+                        snapshot = await client.GetSnapshotAsync(loadOption.SnapshotName).ConfigureAwait(false);
+                    }
+                    catch (RequestFailedException rfe) when (rfe.Status == (int)HttpStatusCode.NotFound)
+                    {
+                        throw new InvalidOperationException($"Could not find snapshot with name '{loadOption.SnapshotName}'.", rfe);
+                    }
+
+                    if (snapshot.SnapshotComposition != SnapshotComposition.Key)
+                    {
+                        throw new InvalidOperationException($"{nameof(snapshot.SnapshotComposition)} for the selected snapshot with name '{snapshot.Name}' must be 'key', found '{snapshot.SnapshotComposition}'.");
+                    }
+
+                    pageableSettings = client.GetConfigurationSettingsForSnapshotAsync(
+                        loadOption.SnapshotName,
+                        cancellationToken);
+                }
 
-                foreach (ConfigurationSetting setting in currentKeyValues)
+                await CallWithRequestTracing(async () =>
                 {
-                    watchedSettings[new KeyValueIdentifier(setting.Key, setting.Label)] = new ConfigurationSetting(setting.Key, setting.Value, setting.Label, setting.ETag);
+                    await foreach (Page<ConfigurationSetting> page in _options.PageableManager.GetPages(pageableSettings).ConfigureAwait(false))
+                    {
+                        foreach (ConfigurationSetting setting in page.Values)
+                        {
+                            keyValueChanges.Add(new KeyValueChange()
+                            {
+                                Key = setting.Key,
+                                Label = setting.Label,
+                                Current = setting,
+                                ChangeType = KeyValueChangeType.Modified,
+                                IsWatchedSetting = false
+                            });
```

Comment:
we can add the line `loadedSettings.Add(setting.Key)` here to avoid another iteration.